### ISSUE TITLE
Build with remote cache "Failed to delete output files after incomplete download"

### ISSUE THREAD

**Issue by bayareabear** [bazelbuild/bazel#6940]
With bazel release 0.20.0, we are seeing "Failed to delete output files after incomplete download" again when set jobs to 2000.

**Comment by buchgr** [comment]
@bayareabear on which platform are you running?

**Comment by bayareabear** [comment]
I am building on Mac OS for iOS app

**Comment by RNabel** [comment]
Can confirm same error on Bazel 2.0.0 on Windows, with following redacted trace:
```
Couldn't build file worker_sdk/package/c-headers.zip: error executing shell command: 'C:/Program Files/Git/usr/bin/bash.exe -c bash "bazel-out/x64_windows-opt-msvc_md/bin/<target path>"' failed due to Failed to delete output files after incomplete download: C:/bazel/2arvwb6u/execroot/__main__/bazel-out/x64_windows-opt-msvc_md/bin/<target path>.tmp (Directory not empty)
--
  | java.io.IOException: C:/bazel/2arvwb6u/execroot/__main__/bazel-out/x64_windows-opt-msvc_md/bin/<target path>.tmp (Directory not empty)
  | at com.google.devtools.build.lib.windows.WindowsFileSystem.delete(WindowsFileSystem.java:60)
  | at com.google.devtools.build.lib.vfs.Path.delete(Path.java:675)
  | at com.google.devtools.build.lib.remote.RemoteCache.download(RemoteCache.java:339)
  | at com.google.devtools.build.lib.remote.RemoteSpawnCache.lookup(RemoteSpawnCache.java:174)
  | at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:117)
  | at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:87)
  | at com.google.devtools.build.lib.actions.SpawnActionContext.beginExecution(SpawnActionContext.java:41)
  | at com.google.devtools.build.lib.exec.ProxySpawnActionContext.beginExecution(ProxySpawnActionContext.java:60)
  | at com.google.devtools.build.lib.analysis.actions.SpawnAction.beginExecution(SpawnAction.java:332)
  | at com.google.devtools.build.lib.actions.Action.execute(Action.java:123)
  | at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$4.execute(SkyframeActionExecutor.java:905)
  | at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.continueAction(SkyframeActionExecutor.java:1040)
  | at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.run(SkyframeActionExecutor.java:1011)
  | at com.google.devtools.build.lib.skyframe.ActionExecutionState.runStateMachine(ActionExecutionState.java:116)
  | at com.google.devtools.build.lib.skyframe.ActionExecutionState.getResultOrDependOnFuture(ActionExecutionState.java:77)
  | at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor.executeAction(SkyframeActionExecutor.java:589)
  | at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.checkCacheAndExecuteIfNeeded(ActionExecutionFunction.java:902)
  | at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.compute(ActionExecutionFunction.java:292)
  | at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:454)
  | at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:399)
  | at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
  | at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
  | at java.base/java.lang.Thread.run(Unknown Source)
  | Suppressed: java.io.FileNotFoundException: C:\bazel\2arvwb6u\execroot\__main__\bazel-out\x64_windows-opt-msvc_md\bin\<target path>.tmp (Access is denied)
  | at java.base/java.io.FileOutputStream.open0(Native Method)
  | at java.base/java.io.FileOutputStream.open(Unknown Source)
  | at java.base/java.io.FileOutputStream.<init>(Unknown Source)
  | at java.base/java.io.FileOutputStream.<init>(Unknown Source)
  | at com.google.devtools.build.lib.vfs.AbstractFileSystem.createFileOutputStream(AbstractFileSystem.java:124)
  | at com.google.devtools.build.lib.vfs.AbstractFileSystem.getOutputStream(AbstractFileSystem.java:131)
  | at com.google.devtools.build.lib.vfs.Path.getOutputStream(Path.java:524)
  | at com.google.devtools.build.lib.vfs.Path.getOutputStream(Path.java:512)
  | at com.google.devtools.build.lib.remote.RemoteCache$LazyFileOutputStream.ensureOpen(RemoteCache.java:1014)
  | at com.google.devtools.build.lib.remote.RemoteCache$LazyFileOutputStream.write(RemoteCache.java:990)
  | at com.google.common.hash.HashingOutputStream.write(HashingOutputStream.java:58)
  | at com.google.devtools.build.lib.remote.http.HttpCacheClient$2.write(HttpCacheClient.java:473)
  | at io.netty.buffer.UnsafeByteBufUtil.getBytes(UnsafeByteBufUtil.java:608)
  | at io.netty.buffer.UnsafeByteBufUtil.getBytes(UnsafeByteBufUtil.java:588)
  | at io.netty.buffer.PooledUnsafeDirectByteBuf.getBytes(PooledUnsafeDirectByteBuf.java:152)
  | at io.netty.buffer.PooledSlicedByteBuf.getBytes(PooledSlicedByteBuf.java:377)
  | at io.netty.buffer.AbstractByteBuf.readBytes(AbstractByteBuf.java:957)
  | at com.google.devtools.build.lib.remote.http.HttpDownloadHandler.channelRead0(HttpDownloadHandler.java:108)
  | at com.google.devtools.build.lib.remote.http.HttpDownloadHandler.channelRead0(HttpDownloadHandler.java:43)
  | at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105)
  | at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359)
  | at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:345)
  | at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:337)
  | at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:438)
  | at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:323)
  | at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:297)
  | at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:253)
  | at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359)
  | at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:345)
  | at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:337)
  | at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1476)
  | at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1225)
  | at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1272)
  | at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:502)
  | at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:441)
  | at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:278)
  | at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359)
  | at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:345)
  | at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:337)
  | at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)
  | at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359)
  | at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:345)
  | at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:337)
  | at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1408)
  | at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359)
  | at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:345)
  | at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:930)
  | at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
  | at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:677)
  | at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:612)
  | at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:529)
  | at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:491)
  | at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:905)
  | at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
  | ... 1 more
  | Suppressed: java.io.FileNotFoundException: C:\bazel\2arvwb6u\execroot\__main__\bazel-out\x64_windows-opt-msvc_md\bin\<target path>.tmp (Access is denied)
  | at java.base/java.io.FileOutputStream.open0(Native Method)
  | at java.base/java.io.FileOutputStream.open(Unknown Source)
  | at java.base/java.io.FileOutputStream.<init>(Unknown Source)
  | at java.base/java.io.FileOutputStream.<init>(Unknown Source)
  | at com.google.devtools.build.lib.vfs.AbstractFileSystem.createFileOutputStream(AbstractFileSystem.java:124)
  | at com.google.devtools.build.lib.vfs.AbstractFileSystem.getOutputStream(AbstractFileSystem.java:131)
  | at com.google.devtools.build.lib.vfs.Path.getOutputStream(Path.java:524)
  | at com.google.devtools.build.lib.vfs.Path.getOutputStream(Path.java:512)
  | at com.google.devtools.build.lib.remote.RemoteCache$LazyFileOutputStream.ensureOpen(RemoteCache.java:1014)
  | at com.google.devtools.build.lib.remote.RemoteCache$LazyFileOutputStream.close(RemoteCache.java:1008)
  | at com.google.devtools.build.lib.remote.RemoteCache$2.onFailure(RemoteCache.java:476)
  | at com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1349)
  | at com.google.common.util.concurrent.MoreExecutors$DirectExecutor.execute(MoreExecutors.java:398)
  | at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1024)
  | at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:866)
  | at com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:711)
  | at com.google.common.util.concurrent.AbstractTransformFuture.run(AbstractTransformFuture.java:93)
  | at com.google.common.util.concurrent.MoreExecutors$DirectExecutor.execute(MoreExecutors.java:398)
  | at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1024)
  | at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:866)
  | at com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:711)
  | at com.google.common.util.concurrent.SettableFuture.setException(SettableFuture.java:54)
  | at com.google.devtools.build.lib.remote.http.HttpCacheClient.lambda$get$5(HttpCacheClient.java:523)
  | at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:502)
  | at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:476)
  | at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:415)
  | at io.netty.util.concurrent.DefaultPromise.setValue0(DefaultPromise.java:540)
  | at io.netty.util.concurrent.DefaultPromise.setFailure0(DefaultPromise.java:533)
  | at io.netty.util.concurrent.DefaultPromise.setFailure(DefaultPromise.java:106)
  | at io.netty.channel.DefaultChannelPromise.setFailure(DefaultChannelPromise.java:89)
  | at com.google.devtools.build.lib.remote.http.AbstractHttpHandler.failAndResetUserPromise(AbstractHttpHandler.java:57)
  | at com.google.devtools.build.lib.remote.http.AbstractHttpHandler.exceptionCaught(AbstractHttpHandler.java:121)
  | at com.google.devtools.build.lib.remote.http.HttpDownloadHandler.exceptionCaught(HttpDownloadHandler.java:158)
  | at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:282)
  | at io.netty.channel.AbstractChannelHandlerContext.notifyHandlerException(AbstractChannelHandlerContext.java:820)
  | at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:361)
  | ... 34 more
  | Caused by: java.nio.file.DirectoryNotEmptyException: C:/bazel/2arvwb6u/execroot/__main__/bazel-out/x64_windows-opt-msvc_md/bin/<target path>.tmp
  | at com.google.devtools.build.lib.windows.jni.WindowsFileOperations.deletePath(WindowsFileOperations.java:254)
  | at com.google.devtools.build.lib.windows.WindowsFileSystem.delete(WindowsFileSystem.java:58)
  | ... 22 more


```
`<target path>` is the same throughout the stack trace.
We set jobs to 100.

@buchgr should this be a separate ticket or could this be related? A similar Windows-specific ticket would be #10363.

**Comment by akharlamov** [comment]
It may be related to the constelation of bug/features that presents in JVM at least since 1.4. 
See Sun bug database for details: 
https://bugs.java.com/bugdatabase/view_bug.do?bug_id=4722539
https://bugs.java.com/bugdatabase/view_bug.do?bug_id=6962459
etc

**TL;DR**
Java zip file implementation caches file handles and keeps them open for some time. On *nix machines it is fine, but windows holds a lock on the open files and prevents deletion.

AFAIR, the only way was to just never move/delete zip files. Copy only.
